atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/VersionedEventStore.java

```diff
@@ -48,11 +50,25 @@
                 .orElseGet(ImmutableList::of);
     }
 
-    LockWatchEvents retentionEvents() {
-        int numToRetention = Math.max(0, eventMap.size() - maxEvents);
+    LockWatchEvents retentionEvents(Optional<Long> earliestSequenceToKeep) {
+        if (eventMap.size() < maxEvents) {
+            return LockWatchEvents.builder().build();
+        }
+
+        int numToRetention = eventMap.size() - maxEvents;
+        long earliestVersion = earliestSequenceToKeep.orElse(Long.MAX_VALUE);
+
         ImmutableLockWatchEvents.Builder builder = LockWatchEvents.builder();
-        Iterators.consumingIterator(Iterators.limit(eventMap.entrySet().iterator(), numToRetention))
-                .forEachRemaining(entry -> builder.addEvents(entry.getValue()));
+        List<Map.Entry<Long, LockWatchEvent>> eventsToClear = eventMap.entrySet().stream()
+                .limit(numToRetention)
+                .filter(entry -> entry.getKey() < earliestVersion)
+                .collect(Collectors.toList());
+
+        eventsToClear.forEach(entry -> {
+            eventMap.remove(entry.getKey(), entry.getValue());
```

Comment:
Is it even possible for this to fail because the value does not match? I would certainly hope not, in which case let's just go with `eventMap.remove(entry.getKey())`?